### PR TITLE
ref(ts) Convert RichListField to typescript

### PR DESCRIPTION
**Pull request by markstory**
I tried to make this work without `any` but the `FieldFromConfig` types for several properties are not compatible with types expected elsewhere. I made some minor changes to the default props as they wouldn't work that well.

### CODE REVIEW

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.tsx [hunk 1]
```diff
@@ -55,11 +62,14 @@ export const fields = {
...
-      builtinSymbolSources
-        .filter(source => !source.hidden)
-        .map(source => [source.sentry_key, t(source.name)]),
+    choices: ({builtinSymbolSources}) => {
+      return (
+        builtinSymbolSources &&
+        ((builtinSymbolSources as BuiltinSymbolSource[])
+          .filter(source => !source.hidden)
+          .map(source => [source.sentry_key, t(source.name)]) as Choices)
+      );
```
Comment (matejminar) [inline src/sentry/static/sentry/app/data/forms/projectDebugFiles.tsx:71]:
```suggestion
      return (builtinSymbolSources as BuiltinSymbolSource[])
        ?.filter(source => !source.hidden)
        .map(source => [source.sentry_key, t(source.name)]) as Choices;
```

**Review by matejminar — APPROVED** [review]

File: src/sentry/static/sentry/app/actionCreators/modal.tsx [hunk 2]
```diff
@@ -189,6 +189,7 @@ export type SentryAppDetailsModalOptions = {
 };
 
 type DebugFileSourceModalOptions = {
+  sourceConfig?: Record<string, string>;
   sourceType: DebugFileSource;
   onSave: (data: Record<string, string>) => void;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/actionCreators/modal.tsx:194]:
```suggestion
  sourceType: DebugFileSource;
  onSave: (data: Record<string, string>) => void;
  sourceConfig?: Record<string, string>;
```

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.tsx [hunk 3]
```diff
@@ -1,126 +1,144 @@
...
-  onRemoveItem: PropTypes.func,
+  addDropdown: DropdownProps;
 
   /**
    * Properties for the confirm remove dialog. If missing, the item will be
    * removed immediately.
    */
-  removeConfirm: PropTypes.object,
-};
+  removeConfirm?: ConfirmProps;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.tsx:85]: can we move this to the end of the object?

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.tsx [hunk 4]
```diff
@@ -194,31 +212,31 @@ class RichList extends React.PureComponent {
...
+    <InputField
+      {...props}
+      field={(fieldProps: RichListProps) => {
+        const {value, ...otherProps} = fieldProps;
+
+        // We must not render this field until `setValue` has been applied by the
+        // model, which is done after the field is mounted for the first time. To
+        // check this, we cannot use Array.isArray because the value passed in by
+        // the model might actually be an ObservableArray.
+        if (typeof value === 'string' || value.length === undefined) {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.tsx:233]: shouldn't we add a `?` in value?
```suggestion
        if (typeof value === 'string' || value?.length === undefined) {
```

**Pull request merged**